Seal indices for faster recovery

brwe opened the pull request:
Add the ability to seal an index. A seal is an id that is written to a primary and its replicas if there are no on going write operations on that shard while the sealing is in progress. Recovery skips file copying for a replica shard if the local copy on a node has the same seal id that the primary has.
Sealing can be triggered in two ways: Automatically, when the IndexingMemoryController detects that shard has become inactive and manually with an api.

relates to #10032

s1monw commented on src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -693,17 +693,28 @@ public CompletionStats completionStats(String... fields) {
+        return engine().syncFlushIfNoPendingChanges(syncId, expectedCommitId);
+    }
+
+    public byte[] flush(FlushRequest request) throws ElasticsearchException {
```
can we add an abstraction to this please? I don't like to return a mutable object as the ID. It can be very simple `CommitID` class that has hashCode & equals etc

s1monw replied:
I also think it should have a `toString` method that prints the base64 representation of the byte array please

brwe replied:
agree, that is much better

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+    /**
+     * send presync requests to all started copies of the given shard
+     */
+    Map<String, byte[]> sendPreSyncRequests(final List<ShardRouting> shards, final ClusterState state, final ShardId shardId) {
```
s/byte[]/CommitID/

s1monw commented on src/main/java/org/elasticsearch/action/admin/indices/seal/SealIndicesRequest.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class SealIndicesRequest extends ActionRequest implements IndicesRequest.Replaceable {
+
+    private String[] indices;
```
@javanna do we have a base class that has all this indices, indicesOptions stuff in it?

javanna replied:
I don't think we have one with just indices and indices options, but maybe we can add it. I see we have BroadcastOperationRequest, maybe we should refactor a bit.

brwe replied:
actually BroadcastOperationRequest fits quite well, no need for refactor. Ill just make SealIndicesRequest inherit from it.

s1monw commented on src/main/java/org/elasticsearch/action/admin/indices/seal/SealIndicesResponse.java:
```diff
@@ -0,0 +1,122 @@
+ */
+public class SealIndicesResponse extends ActionResponse implements ToXContent {
+
+    private Set<SyncedFlushService.SyncedFlushResult> results;
```
can we make it final and assign it in the default ctor... in readFrom we can just call `results.clear()`

s1monw commented on src/main/java/org/elasticsearch/action/admin/indices/seal/TransportSealIndicesAction.java:
```diff
@@ -0,0 +1,93 @@
+        GroupShardsIterator primaries = state.routingTable().activePrimaryShardsGrouped(concreteIndices, false);
+        final Set<SyncedFlushService.SyncedFlushResult> results = ConcurrentCollections.newConcurrentSet();
+        if (primaries.size() == 0) {
+            // no active primary available
```
should we signal this to the user?

brwe replied:
I changed so that whenever a primary is not available a response with error message "no active primary available" is added

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -331,17 +331,27 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
-                                    lastSizeMatched = sizeMatched;
-                                    lastDiscoNodeMatched = discoNode;
+                                // see if we have a sync id we can make use of
+                                if (storeFilesMetaData.syncId() != null && storeFilesMetaData.syncId().equals(primaryNodeStore.syncId())) {
```
can we for better readabilty assign the syncIDs to local vars ie `primarySyncId` and `replicaSyncId`

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -331,17 +331,27 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
+                                } else {
+                                    for (StoreFileMetaData storeFileMetaData : storeFilesMetaData) {
+                                        logger.trace("{}: node [{}] has file {}",
+                                                shard, discoNode.name(), storeFileMetaData.name());
```
do me a favor and assign `storeFileMetaData.name()` to a local var

s1monw commented on src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -218,6 +220,21 @@ public Condition newCondition() {
+
+    public enum SyncedFlushResult {
+        SUCCESS,
+        FAILED_COMMIT_MISMATCH,
```
I think we can remove the `FAILED_` ? `COMMIT_MISMATCH` is fine?

s1monw commented on src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -218,6 +220,21 @@ public Condition newCondition() {
+     * @param expectedCommitId the expected value of
+     * @return true if the sync commit was made, false o.w.
+     */
+    public abstract SyncedFlushResult syncFlushIfNoPendingChanges(String syncId, byte[] expectedCommitId) throws EngineException;
```
I don't like the fact that we have an impl detail in the method name maybe we have a way to syncFlush if there are pending changes? just call it `syncFlush`?

s1monw commented on src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -439,16 +456,19 @@ public final boolean refreshNeeded() {
+     * @return the commit Id for the resulting commit
      */
-    public abstract void flush(boolean force, boolean waitIfOngoing) throws EngineException;
+    public abstract byte[] flush(boolean force, boolean waitIfOngoing) throws EngineException;
```
yeah this should really return a `CommitID` Object

s1monw commented on src/main/java/org/elasticsearch/indices/InternalIndicesLifecycle.java:
```diff
@@ -245,4 +245,15 @@ public void indexShardStateChanged(IndexShard indexShard, @Nullable IndexShardSt
             }
         }
     }
+
```
add missing `@Override` annotation

nik9000 replied:
Why not use checkstyle or something to catch those?

brwe replied:
It does not override any method from the superclass. onShardInactive() is a method in IndicesLifecycle.Listener.

s1monw replied:
ah yeah sorry

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+    * Tries to flush all copies of a shard and write a sync id to it.
+    * After a synced flush two shard copies may only contain the same sync id if they contain the same documents.
+    * To ensure this, synced flush works in three steps:
+    * 1. Flush all shard copies and gather the commit points for each copy after the flush
```
s/commit point/commit ID/

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+    * Step 3 will be executed on primary and replica as well because there are no uncommitted changes on primary (the first flush committed them) and there are no uncommitted
+    * changes on the replica (the write operation has not reached the replica yet).
+    *
+    * Step 2 detects this scenario and fails the whole synced flush if a write operation is ongoing on the primary.
```
can we add here that the neat part about this check in step 2. is that it will fail the flush on the primary even if there is an in-flight request coming into the primary right in between step 2. and 3. since  it will not execute the sync flush if there is a single modification in the buffer of these shards. I think that must be more clear here.

brwe replied:
I added a few more lines but not sure if this is what you meant...

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+            }
+
+            int inflight = getInflightOpsCount(shardId, state, shardRoutingTable);
+            if (inflight != 1) {
```
can we add a comment why we are checking for `1` here?

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+                actionListener.onResponse(new SyncedFlushResult(shardId, "all shards failed to commit on pre-sync"));
+            }
+
+            int inflight = getInflightOpsCount(shardId, state, shardRoutingTable);
```
can we add an assertion that `inflight >= -1` just for kicks

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+            if (latch.await(inflightOpsTimeout.millis(), TimeUnit.MILLISECONDS) == false) {
+                logger.debug("{} in flight operation check timed out after [{}]", shardId, syncTimeout);
+            }
+        } catch (InterruptedException e) {
```
we have to restore the interrupt status here with `Thread.currentThread.interrupt()`

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+
+
+    void sendSyncRequests(final String syncId, List<ShardRouting> shards, ClusterState state, Map<String, byte[]> expectedCommitIds, final ShardId shardId, final ActionListener<SyncedFlushResult> listener) {
+        final CountDown countDownLatch = new CountDown(shards.size());
```
s/countDownLatch/countDown

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+                if (countDownLatch.countDown()) {
+                    listener.onResponse(new SyncedFlushResult(shardId, syncId, results));
+                }
+                continue;
```
if we call the listener we should be done here right, can we assert this somehow? ie `assert shard.equals(shards.get(shards.size()-1);`

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+                logger.trace("{} can't resolve expected commit id for {}, skipping for sync id [{}]. shard routing {}", shardId, syncId, shard);
+                results.put(shard, new SyncedFlushResponse("no commit id from pre-sync flush"));
+                if (countDownLatch.countDown()) {
+                    listener.onResponse(new SyncedFlushResult(shardId, syncId, results));
```
see comment above about asserting

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+
+                @Override
+                public void handleResponse(PreSyncedFlushResponse response) {
+                    byte[] existing = commitIds.put(node.id(), response.commitId());
```
if you make `commitIds` a `ConcurrentMap` you can use `putIfAbsent` here that might be better?

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+                @Override
+                public void handleResponse(PreSyncedFlushResponse response) {
+                    byte[] existing = commitIds.put(node.id(), response.commitId());
+                    assert existing == null : "got two answers for node [" + node + "]";
```
put the assert after the `countDown` to ensure we are not deadlocking?

brwe replied:
actually, we have to put the assert before. In case it fails we call handleFailure and therefore would decrement twice if we first decrement and then assert (although I really think if this assertion fails we have a much worse problem.)
There is a comment in the code just beneath this line.

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+            if (countDownLatch.await(preSyncTimeout.millis(), TimeUnit.MILLISECONDS) == false) {
+                logger.debug("{} waiting for pre sync flush requests timed out after [{}]. pending ops [{}]", shardId, preSyncTimeout, countDownLatch.getCount());
+            }
+        } catch (InterruptedException e) {
```
restore interrupt status

s1monw commented on src/main/java/org/elasticsearch/indices/SyncedFlushService.java:
```diff
@@ -0,0 +1,755 @@
+
+    final static class PreSyncedFlushRequest extends TransportRequest {
+        private ShardId shardId;
+
```
extra newline

s1monw commented on src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -201,28 +202,45 @@ public void phase1(final SnapshotIndexCommit snapshot, final Translog.View trans
+                    recoverySourceSyncId.equals(recoveryTargetSyncId);
+            if (recoverWithSyncId) {
+                for (StoreFileMetaData md : request.metadataSnapshot()) {
+                    response.phase1ExistingFileNames.add(md.name());
```
hmm this is not necessarily correct no? I there files are not existing at all some might but we are really just skipping this phase... I don't know if it's a big deal but we should revisit this and maybe have a `equivalentFileNames` list or so? WDYT

s1monw replied:
when we recover by SyncID can we assert that both shards have the same amount of live documents ie. IndexReader.numDocs() must be the same?

brwe replied:
For the first point: maybe we can just rename phase1ExistingFileNames to phase1ExistingOrEquivalentFileNames? We do not want to change anything in the behavior just the naming is confusing, right?

For the second I will probably need your help. Unclear to me a which state we can actually check that.

s1monw replied:
@brwe I looked into this again and we discussed this briefly before. I think we should add this

```Java
  /**
     * Returns the number of document in the index referenced by this {@link SegmentInfos}
     */
    public static int getNumDocs(SegmentInfos info) {
        int numDocs = 0;
        for (SegmentCommitInfo si : info) {
            numDocs +=si.info.maxDoc() - si.getDelCount();
        }
        return numDocs;
    }
```

to `Lucene.java` as a util and add the value to `Store.MetadataSnapshot` and transfer it with our request before we do the phase 1 comparison. That way we can simply compare the number and fail everything if they don't match?

s1monw commented on src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -398,8 +415,9 @@ public void run() throws InterruptedException {
                     // related to this recovery (out of date segments, for example)
                     // are deleted
                     try {
+                        final Store.MetadataSnapshot remainingFilesAfterCleanup = recoverWithSyncId? request.metadataSnapshot(): recoverySourceMetadata;
```
s/recoverWithSyncId?/recoverWithSyncId ?/

s1monw commented:
I added a bunch of comments nothing really major, looks pretty good though! I wonder if we can add some more assertions to the after test stuff and fetch the sync IDs and if two shards have the same sync ID they also need to have the same amount of committed and live documents?

nik9000 commented on docs/reference/indices/seal.asciidoc:
```diff
@@ -0,0 +1,74 @@
+--------------------------------------------------
+
+The response contains details about for which shards a seal was written and the reason in case of failure.
+Response in case all copies of a shard successfully wrote the seal:
```
Can this return a non-200 result if there were errors sealing the index? I took a glance and it doesn't look like that is what it does now.

brwe replied:
yes, that makes sense. not exactly an expert in rest statuses though. Do you think it would be ok to return CONFLICT in case the operation failed? Also in case it only succeeded on some shards should we return PARTIAL_CONTENT or something of the likes?

clintongormley replied:
I think for other similar APIs (eg delete index) that we only return an exception if the whole request failed, otherwise we return success (and detail failures in the body... I think...).  We definitely shouldn't use CONFLICT - that is used just for document version conflicts, and I don't think we use PARTIAL_CONTENT anywhere at the moment.

@javanna what behaviour would be consistent with related APIs?

nik9000 replied:
I think it'd be surprising to get a 200 response from _seal if it didn't fully seal. Folks won't check for the response if they see the 200 - they will just expect everything to have worked.

I'm not super sure what the right response should be. I suspect it depends on if you consider being unable to fully seal an index to be the server's fault ([5xx](http://tools.ietf.org/html/rfc7231#section-6.6)) or the client's ([4xx](http://tools.ietf.org/html/rfc7231#section-6.5). I _think_ its the client's fault for sending the request while some other client is sending another index request. From what I can see (conflict)[http://tools.ietf.org/html/rfc7231#section-6.5.8] might be the closest because it implies that the client might be able to resolve the issue. _seal isn't a resource in the same way that a document is but its the closest thing I see. Better than 418.

brwe commented:
@nik9000 @s1monw thanks for the review! I addressed all comments or answered except for the recovery part here https://github.com/elastic/elasticsearch/pull/11179#discussion_r30437383 where I still have to dig how this can be done. 
I also added another check after each test to see if shards with same sync id also have same documents. 
Other than that I am unsure about the rest statuses that should be returned for the rest api. I now return CONFLICT(409) if al failed and PARTIAL_CONTENT(206) if only some shards succeeded but that seems odd to me since it is not about content. What would be appropriate here?

nik9000 commented on docs/reference/indices/seal.asciidoc:
```diff
@@ -0,0 +1,74 @@
+The seal API allows to flush one or more indices and adds a marker to primaries and replicas if there are no pending write operations.
+The seal marker is used during recovery after a node restarts. If a replica is allocated on a node which already has a shard copy with the same seal as the primary then no files will be copied during recovery.
+Sealing is a best effort operation. If write operations are ongoing while the sealing is in progress then writing the seal might fail on some copies.
+
```
Is it worth adding information here about what happens if a write operation comes after the _seal/how to unseal the index? I suspect the seal is paper thin - the write operation just obliterates it.

s1monw replied:
@brwe I left an example up there for how to get the numDocs

brwe replied:
done, that was actually easy..

@nik9000 that is totally worth adding, I'll do that

brwe replied:
actually @s1monw did now :)

s1monw replied:
@nik9000 we added a paragraph - wanna take another look?

nik9000 replied:
I was trying to comment on it but it was easier to just rewrite some of it and make assumptions then to ask questions. So I sent #11235. Is it right?

s1monw commented:
I think we are ready here LGTM @brwe
